protocol/directory.go

```diff
@@ -0,0 +1,48 @@
+package protocol
+
+import (
+	"log"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type ConiksUserPolicies struct {
+	Username               string
+	AllowUnsignedKeychange bool
+	AllowPublicLookup      bool
+}
+
+type ConiksDirectory struct {
+	*merkletree.PAD
+}
+
+func InitDirectory(policies merkletree.Policies, signKey sign.PrivateKey,
+	dirSize uint64) *ConiksDirectory {
+	pad, err := merkletree.NewPAD(policies, signKey, dirSize)
+	if err != nil {
+		panic(err)
+	}
+	return &ConiksDirectory{pad}
+}
+
+func (d *ConiksDirectory) Register(uname string, key []byte) (
+	*merkletree.AuthenticationPath, *merkletree.TemporaryBinding, ErrorCode) {
+	// check whether the name already exists
+	// in the directory before we register
+	ap, err := d.Lookup(uname)
+	if err != nil {
+		return nil, nil, ErrorInternalServer
+	}
+	if ap.Leaf.Value() != nil {
```

Comment:
Ok, so all this time you've been right and I'm trying to explain something to you that *I* don't understand.  That's kind of sad.

However, judging by how long it took me to catch on, I would say two things,

 * Checking the index is always the correct thing to do here, whereas value == nil feels like an implementation detail, in that someone could choose to suppress the value at another location, like serialization time. So, I think we should leave the LookupIndex == Index() comparison you implemented.

* Now that I see what's going, I absolutely agree that we should merge eb7245df72cebc2de19f43e042d4d68314262b96.

Finally, I can't believe I said,

> Please think about it some more.

I'm going to go hide in shame.